# Exercise 15

# Fraud Detection 

## Introduction

- Fraud Detection Dataset from Microsoft Azure: [data](http://gallery.cortanaintelligence.com/Experiment/8e9fe4e03b8b4c65b9ca947c72b8e463)

Fraud detection is one of the earliest industrial applications of data mining and machine learning. Fraud detection is typically handled as a binary classification problem, but the class population is unbalanced because instances of fraud are usually very rare compared to the overall volume of transactions. Moreover, when fraudulent transactions are discovered, the business typically takes measures to block the accounts from transacting to prevent further losses. 

In [42]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [43]:
import pandas as pd

url = 'https://raw.githubusercontent.com/albahnsen/PracticalMachineLearningClass/master/datasets/15_fraud_detection.csv.zip'
df = pd.read_csv(url, index_col=0)
df.head()

accountAge  digitalItemCount  sumPurchaseCount1Day  sumPurchaseAmount1Day  \
0        2000                 0                     0                   0.00   
1          62                 1                     1                1185.44   
2        2000                 0                     0                   0.00   
3           1                 1                     0                   0.00   
4           1                 1                     0                   0.00   

   sumPurchaseAmount30Day  paymentBillingPostalCode - LogOddsForClass_0  \
0                  720.25                                      5.064533   
1                 2530.37                                      0.538996   
2                    0.00                                      5.064533   
3                    0.00                                      5.064533   
4                  132.73                                      5.412885   

   accountPostalCode - LogOddsForClass_0  \
0                               0.421214   
1                               0.481838   
2                               5.096396   
3                               5.096396   
4                               0.342945   

   paymentBillingState - LogOddsForClass_0  accountState - LogOddsForClass_0  \
0                                 1.312186                          0.566395   
1                                 4.401370                          4.500157   
2                                 3.056357                          3.155226   
3                                 3.331154                          3.331239   
4                                 5.563677                          4.086965   

   paymentInstrumentAgeInAccount  ipState - LogOddsForClass_0  \
0                    3279.574306                     1.218157   
1                      61.970139                     4.035601   
2                       0.000000                     3.314186   
3                       0.000000                     3.529398   
4                       0.001389                     3.529398   

   transactionAmount  transactionAmountUSD  ipPostalCode - LogOddsForClass_0  \
0             599.00            626.164650                          1.259543   
1            1185.44           1185.440000                          3.981118   
2              32.09             32.090000                          5.008490   
3             133.28            132.729554                          1.324925   
4             543.66            543.660000                          2.693451   

   localHour - LogOddsForClass_0  Label  
0                       4.745402      0  
1                       4.921349      0  
2                       4.742303      0  
3                       4.745402      0  
4                       4.876771      0

In [44]:
df.shape, df.Label.sum(), df.Label.mean()

((138721, 16), 797, 0.0057453449730033666)

# Exercise 15.1

Estimate a Logistic Regression and a Decision Tree

Evaluate using the following metrics:
* Accuracy
* F1-Score
* F_Beta-Score (Beta=10)

Comment about the results

In [45]:
df['Fraud']=df['Label']==1
df.head()

accountAge  digitalItemCount  sumPurchaseCount1Day  sumPurchaseAmount1Day  \
0        2000                 0                     0                   0.00   
1          62                 1                     1                1185.44   
2        2000                 0                     0                   0.00   
3           1                 1                     0                   0.00   
4           1                 1                     0                   0.00   

   sumPurchaseAmount30Day  paymentBillingPostalCode - LogOddsForClass_0  \
0                  720.25                                      5.064533   
1                 2530.37                                      0.538996   
2                    0.00                                      5.064533   
3                    0.00                                      5.064533   
4                  132.73                                      5.412885   

   accountPostalCode - LogOddsForClass_0  \
0                               0.421214   
1                               0.481838   
2                               5.096396   
3                               5.096396   
4                               0.342945   

   paymentBillingState - LogOddsForClass_0  accountState - LogOddsForClass_0  \
0                                 1.312186                          0.566395   
1                                 4.401370                          4.500157   
2                                 3.056357                          3.155226   
3                                 3.331154                          3.331239   
4                                 5.563677                          4.086965   

   paymentInstrumentAgeInAccount  ipState - LogOddsForClass_0  \
0                    3279.574306                     1.218157   
1                      61.970139                     4.035601   
2                       0.000000                     3.314186   
3                       0.000000                     3.529398   
4                       0.001389                     3.529398   

   transactionAmount  transactionAmountUSD  ipPostalCode - LogOddsForClass_0  \
0             599.00            626.164650                          1.259543   
1            1185.44           1185.440000                          3.981118   
2              32.09             32.090000                          5.008490   
3             133.28            132.729554                          1.324925   
4             543.66            543.660000                          2.693451   

   localHour - LogOddsForClass_0  Label  Fraud  
0                       4.745402      0  False  
1                       4.921349      0  False  
2                       4.742303      0  False  
3                       4.745402      0  False  
4                       4.876771      0  False

In [46]:
y = df['Fraud'].values 
X= df.drop(['Fraud','transactionAmount','Label'], axis=1)

In [47]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
## Estandarizamos los valores de las variables continuas
scaler  =  StandardScaler ()

X_train_ = scaler.fit_transform(X_train)
X_test_ = scaler.transform (X_test)

In [48]:
logreg = LogisticRegression(solver='liblinear',C=1e9,random_state=42)
logreg.fit(X_train_, y_train)  

LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='ovr', n_jobs=1, penalty='l2', random_state=42,
          solver='liblinear', tol=0.0001, verbose=0, warm_start=False)

In [49]:
y_pred_log = logreg.predict(X_test_) 

In [50]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import fbeta_score

In [51]:
accuracy_logistic=accuracy_score(y_test, y_pred_log)
f1_logistic=f1_score(y_test,y_pred_log,average='binary')
fbs_logistic=fbeta_score(y_test,y_pred_log,average='binary', beta = 10 )

print("Accuracy score of a logistic regression:",accuracy_logistic)
print("f1 score of a logistic regression:", f1_logistic)
print("fbeta score of a logistic regression:", fbs_logistic)


Accuracy score of a logistic regression: 0.994112982675349
f1 score of a logistic regression: 0.0
fbeta score of a logistic regression: 0.0


C:\Anaconda\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


In [52]:
tree = DecisionTreeClassifier(max_depth=4,max_leaf_nodes=10,random_state=42)
tree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=4,
            max_features=None, max_leaf_nodes=10,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')

In [53]:
y_pred_tree = tree.predict(X_test)

In [54]:
accuracy_tree=accuracy_score(y_test, y_pred_tree)
f1_tree=f1_score(y_test,y_pred_tree,average='binary')
fbs_tree=fbeta_score(y_test,y_pred_tree,average='binary', beta = 10 )

print("Accuracy score of a tree:",accuracy_tree)
print("f1 score of a tree:", f1_tree)
print("fbeta score of a tree:", fbs_tree)

Accuracy score of a tree: 0.9940649253910662
f1 score of a tree: 0.03891050583657587
fbeta score of a tree: 0.020602154046997386


In [55]:
n_samples = df.shape[0]
print('observaciones ',n_samples)
n_samples_0 = (df['Label'] == 0).sum()
print('Casos sin fraude ',n_samples_0)
n_samples_1 = (df['Label'] == 1).sum()
print('Casos con fraude ',n_samples_1)

observaciones  138721
Casos sin fraude  137924
Casos con fraude  797


El alto accuracy score y el bajo f1 score son los esperados debido a que la base está desbalanceada, tal como se mostró en la línea de código anterior. Por lo tanto, es necesario encontrar una forma de balancear la base de datos, ya sea utilizando over-sampling o under-sampling

# Exercise 15.2

Under-sample the negative class using random-under-sampling

Which is parameter for target_percentage did you choose?
How the results change?

**Only apply under-sampling to the training set, evaluate using the whole test set**

In [56]:
def UnderSampling(X, y, target_percentage=0.5, seed=None):
    # Assuming minority class is the positive
    n_samples = y.shape[0]
    n_samples_0 = (y == 0).sum()
    n_samples_1 = (y == 1).sum()

    n_samples_0_new =  n_samples_1 / target_percentage - n_samples_1
    n_samples_0_new_per = n_samples_0_new / n_samples_0

    filter_ = y == 0

    np.random.seed(seed)
    rand_1 = np.random.binomial(n=1, p=n_samples_0_new_per, size=n_samples)
    
    filter_ = filter_ & rand_1
    filter_ = filter_ | (y == 1)
    filter_ = filter_.astype(bool)
    
    return X[filter_], y[filter_]

In [57]:
for percentage in [0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5]:
    X_porct, y_porct = UnderSampling(X_train_, y_train, percentage, 42)
    logreg.fit(X_porct, y_porct)
    y_pred_log = logreg.predict(X_test_)
    accuracy_logistic1=accuracy_score(y_test, y_pred_log)
    f1_logistic1=f1_score(y_test,y_pred_log,average='binary')
    fbs_logistic1=fbeta_score(y_test,y_pred_log,average='binary', beta = 10 )
    print('Target percentage:',percentage)
    print("Accuracy score of a logistic regression:",accuracy_logistic1)
    print("f1 score of a logistic regression:", f1_logistic1)
    print("fbeta score of a logistic regression:", fbs_logistic1)
    print('\n')


Target percentage: 0.1
Accuracy score of a logistic regression: 0.9883220799192638
f1 score of a logistic regression: 0.04330708661417323
fbeta score of a logistic regression: 0.04486532326454791


Target percentage: 0.15
Accuracy score of a logistic regression: 0.9788547949155393
f1 score of a logistic regression: 0.05172413793103448
fbeta score of a logistic regression: 0.09625541039590199


Target percentage: 0.2
Accuracy score of a logistic regression: 0.9664800442127015
f1 score of a logistic regression: 0.049079754601226995
fbeta score of a logistic regression: 0.14135759272218337


Target percentage: 0.25
Accuracy score of a logistic regression: 0.9442295215897349
f1 score of a logistic regression: 0.04209657449442839
fbeta score of a logistic regression: 0.19308044081265466


Target percentage: 0.3
Accuracy score of a logistic regression: 0.9148184636086215
f1 score of a logistic regression: 0.03642294101658059
fbeta score of a logistic regression: 0.24224958831531465


Target 

In [58]:
for percentage in [0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5]:
    X_porct, y_porct = UnderSampling(X_train, y_train, percentage, 42)
    tree.fit(X_porct, y_porct)
    y_pred_tree = tree.predict(X_test)
    accuracy_tree1=accuracy_score(y_test, y_pred_tree)
    f1_tree1=f1_score(y_test,y_pred_tree,average='binary')
    fbs_tree1=fbeta_score(y_test,y_pred_tree,average='binary', beta = 10 )
    print('Target percentage:',percentage)
    print("Accuracy score of a tree:",accuracy_tree1)
    print("f1 score of a tree:", f1_tree1)
    print("fbeta score of a tree:", fbs_tree1)
    print('\n')

Target percentage: 0.1
Accuracy score of a tree: 0.977389047744912
f1 score of a tree: 0.12465116279069767
fbeta score of a tree: 0.2671535728385314


Target percentage: 0.15
Accuracy score of a tree: 0.9740490664872528
f1 score of a tree: 0.09849749582637729
fbeta score of a tree: 0.2341177857227046


Target percentage: 0.2
Accuracy score of a tree: 0.9505009971886489
f1 score of a tree: 0.06955736224028906
fbeta score of a tree: 0.2938154067021799


Target percentage: 0.25
Accuracy score of a tree: 0.941562342312036
f1 score of a tree: 0.0681992337164751
fbeta score of a tree: 0.33459892052856877


Target percentage: 0.3
Accuracy score of a tree: 0.9270250138164692
f1 score of a tree: 0.06352143077397471
fbeta score of a tree: 0.37831842315804787


Target percentage: 0.35
Accuracy score of a tree: 0.8940096595141409
f1 score of a tree: 0.054853224769659314
fbeta score of a tree: 0.4469953668487657


Target percentage: 0.4
Accuracy score of a tree: 0.8964365523704255
f1 score of a tre

El valor del porcentaje de balanceo adecuado es aquel que reporta el f1 score más alto. Para ambos casos, estos porcentajes son 10% y 15%.Los resultados de rendimiento mejoraron comparados con aquellos obtenidos en el ejercicio 15.1

# Exercise 15.3

Same analysis using random-over-sampling

In [59]:
def OverSampling(X, y, target_percentage=0.5, seed=None):
    # Assuming minority class is the positive
    n_samples = y.shape[0]
    n_samples_0 = (y == 0).sum()
    n_samples_1 = (y == 1).sum()

    n_samples_1_new =  -target_percentage * n_samples_0 / (target_percentage- 1)

    np.random.seed(seed)
    filter_ = np.random.choice(X[y == 1].shape[0], int(n_samples_1_new))
    # filter_ is within the positives, change to be of all
    filter_ = np.nonzero(y == 1)[0][filter_]
    
    filter_ = np.concatenate((filter_, np.nonzero(y == 0)[0]), axis=0)
    
    return X[filter_], y[filter_]

In [60]:
for percentage in [0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5]:
    X_porct, y_porct = OverSampling(X_train_, y_train, percentage, 42)
    logreg.fit(X_porct, y_porct)
    y_pred_log = logreg.predict(X_test_)
    accuracy_logistic2=accuracy_score(y_test, y_pred_log)
    f1_logistic2=f1_score(y_test,y_pred_log,average='binary')
    fbs_logistic2=fbeta_score(y_test,y_pred_log,average='binary', beta = 10 )
    print('Target percentage:',percentage)
    print("Accuracy score of a logistic regression:",accuracy_logistic2)
    print("f1 score of a logistic regression:", f1_logistic2)
    print("fbeta score of a logistic regression:", fbs_logistic2)
    print('\n')

Target percentage: 0.1
Accuracy score of a logistic regression: 0.9914698320397914
f1 score of a logistic regression: 0.037940379403794036
fbeta score of a logistic regression: 0.028711825860948666


Target percentage: 0.15
Accuracy score of a logistic regression: 0.9862315880529591
f1 score of a logistic regression: 0.04975124378109452
fbeta score of a logistic regression: 0.06094617426985274


Target percentage: 0.2
Accuracy score of a logistic regression: 0.9778696205877406
f1 score of a logistic regression: 0.051493305870236865
fbeta score of a logistic regression: 0.10009513993498773


Target percentage: 0.25
Accuracy score of a logistic regression: 0.9658072422327414
f1 score of a logistic regression: 0.04560697518443998
fbeta score of a logistic regression: 0.1333799425153422


Target percentage: 0.3
Accuracy score of a logistic regression: 0.9468726722252926
f1 score of a logistic regression: 0.04409857328145266
fbeta score of a logistic regression: 0.1938798554652213


Target 

In [61]:
for percentage in [0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5]:
    X_porct, y_porct = OverSampling(X_train_, y_train, percentage, 42)
    tree.fit(X_porct, y_porct)
    y_pred_tree = tree.predict(X_test_)
    accuracy_tree2=accuracy_score(y_test, y_pred_tree)
    f1_tree2=f1_score(y_test,y_pred_tree,average='binary')
    fbs_tree2=fbeta_score(y_test,y_pred_tree,average='binary', beta = 10 )
    print('Target percentage:',percentage)
    print("Accuracy score of a tree:",accuracy_tree2)
    print("f1 score of a tree:", f1_tree2)
    print("fbeta score of a tree:", fbs_tree2)
    print('\n')    

Target percentage: 0.1
Accuracy score of a tree: 0.9846216690294831
f1 score of a tree: 0.16010498687664043
fbeta score of a tree: 0.24627253467642002


Target percentage: 0.15
Accuracy score of a tree: 0.9839248384073817
f1 score of a tree: 0.12549019607843137
fbeta score of a tree: 0.1937649880095923


Target percentage: 0.2
Accuracy score of a tree: 0.969723910901795
f1 score of a tree: 0.08959537572254335
fbeta score of a tree: 0.24423729474628494


Target percentage: 0.25
Accuracy score of a tree: 0.9369488430208809
f1 score of a tree: 0.06619217081850533
fbeta score of a tree: 0.3470533899870682


Target percentage: 0.3
Accuracy score of a tree: 0.9373092726530023
f1 score of a tree: 0.061848256023013304
fbeta score of a tree: 0.3212753365882527


Target percentage: 0.35
Accuracy score of a tree: 0.8264651464545738
f1 score of a tree: 0.04039330321551953
fbeta score of a tree: 0.4830559139108272


Target percentage: 0.4
Accuracy score of a tree: 0.8264651464545738
f1 score of a t

En este caso el porcentaje adecuado es 20% cuando se corre una regresión logística y 10% cuando se utiliza un árbol. Al igual que el numeral 15.2, los resultados en términos del f1 score son mejores comparados con aquellos del ejercicio 15.1

# Exercise 15.4 (3 points)

Evaluate the results using SMOTE

Which parameters did you choose?

In [65]:
from imblearn.over_sampling import SMOTE

In [66]:
for k in [5,6,7,8,9,10]: # número de vecinos más cercanos
    print(k,' vecinos:')
    for percentage in [0.1,0.15,0.2,0.3,0.4,0.5]:
        smote = SMOTE(percentage,random_state=42, k_neighbors = k)
        X_porct, y_porct = smote.fit_resample(X_train_, y_train)
        logreg.fit(X_porct, y_porct)
        y_pred_log = logreg.predict(X_test_)
        accuracy_logistic3=accuracy_score(y_test, y_pred_log)
        f1_logistic3=f1_score(y_test,y_pred_log,average='binary')
        fbs_logistic3=fbeta_score(y_test,y_pred_log,average='binary', beta = 10 )
        print('Target percentage:',percentage)
        print("Accuracy score of a logistic regression:",accuracy_logistic3)
        print("f1 score of a logistic regression:", f1_logistic3)
        print("fbeta score of a logistic regression:", fbs_logistic3)
        print('\n')

5  vecinos:
Target percentage: 0.1
Accuracy score of a logistic regression: 0.9924790350097316
f1 score of a logistic regression: 0.024922118380062308
fbeta score of a logistic regression: 0.016438802083333336


Target percentage: 0.15
Accuracy score of a logistic regression: 0.9894273974577696
f1 score of a logistic regression: 0.03930131004366813
fbeta score of a logistic regression: 0.03678226034880428


Target percentage: 0.2
Accuracy score of a logistic regression: 0.984885984093039
f1 score of a logistic regression: 0.05697151424287856
fbeta score of a logistic regression: 0.07700024075114358


Target percentage: 0.3
Accuracy score of a logistic regression: 0.9719585746209481
f1 score of a logistic regression: 0.044226044226044224
fbeta score of a logistic regression: 0.1070419218087612


Target percentage: 0.4
Accuracy score of a logistic regression: 0.9527116322656607
f1 score of a logistic regression: 0.04651162790697675
fbeta score of a logistic regression: 0.1842015274136555

En este caso el fbeta score se maximiza (0.2388) cuando se utilizan 8 vecinos más cercanos y un porcentaje de balanceo igual al 50%.

In [67]:
for k in [5,6,7,8,9,10]: #número de vecinos
    print(k,' vecinos:')
    for percentage in [0.1,0.15,0.2,0.3,0.4,0.5]:
        smote = SMOTE(percentage,random_state=42, k_neighbors = k)
        X_porct, y_porct = smote.fit_resample(X_train_, y_train)
        tree.fit(X_porct, y_porct)
        y_pred_tree = tree.predict(X_test_)
        accuracy_tree3=accuracy_score(y_test, y_pred_tree)
        f1_tree3=f1_score(y_test,y_pred_tree,average='binary')
        fbs_tree3=fbeta_score(y_test,y_pred_tree,average='binary', beta = 10 )
        print('Target percentage:',percentage)
        print("Accuracy score of a tree:",accuracy_tree3)
        print("f1 score of a tree:", f1_tree3)
        print("fbeta score of a tree:", fbs_tree3)
        print('\n')    

5  vecinos:
Target percentage: 0.1
Accuracy score of a tree: 0.9919023475983372
f1 score of a tree: 0.07162534435261707
fbeta score of a tree: 0.05333495816069543


Target percentage: 0.15
Accuracy score of a tree: 0.9864238171900905
f1 score of a tree: 0.13740458015267176
fbeta score of a tree: 0.18245684464070652


Target percentage: 0.2
Accuracy score of a tree: 0.9863997885479492
f1 score of a tree: 0.13719512195121952
fbeta score of a tree: 0.18244952029224037


Target percentage: 0.3
Accuracy score of a tree: 0.9611456856573035
f1 score of a tree: 0.05714285714285715
fbeta score of a tree: 0.19056603773584904


Target percentage: 0.4
Accuracy score of a tree: 0.9410337121849244
f1 score of a tree: 0.061208875286916604
fbeta score of a tree: 0.3007182998995125


Target percentage: 0.5
Accuracy score of a tree: 0.8152918278588077
f1 score of a tree: 0.03417514763161202
fbeta score of a tree: 0.4263984602967653


6  vecinos:
Target percentage: 0.1
Accuracy score of a tree: 0.9874089

En este caso el fbeta score se maximiza (0.42) cuando se utilizan 5 vecinos más cercanos y un porcentaje de balanceo igual al 50%.

# Exercise 15.5 (3 points)

Evaluate the results using Adaptive Synthetic Sampling Approach for Imbalanced
Learning (ADASYN)

http://www.ele.uri.edu/faculty/he/PDFfiles/adasyn.pdf
https://imbalanced-learn.readthedocs.io/en/stable/generated/imblearn.over_sampling.ADASYN.html#rf9172e970ca5-1

In [68]:
from imblearn.over_sampling import ADASYN

In [70]:
for k in [5,6,7,8,9,10]: # número de vecinos más cercanos
    print(k,' vecinos:')
    for percentage in [0.1,0.15,0.2,0.3,0.4,0.5]:
        adasyn = ADASYN(percentage,random_state=42, n_neighbors = k)
        X_porct, y_porct = adasyn.fit_resample(X_train_, y_train)
        logreg.fit(X_porct, y_porct)
        y_pred_log = logreg.predict(X_test_)
        accuracy_logistic4=accuracy_score(y_test, y_pred_log)
        f1_logistic4=f1_score(y_test,y_pred_log,average='binary')
        fbs_logistic4=fbeta_score(y_test,y_pred_log,average='binary', beta = 10 )
        print('Target percentage:',percentage)
        print("Accuracy score of a logistic regression:",accuracy_logistic4)
        print("f1 score of a logistic regression:", f1_logistic4)
        print("fbeta score of a logistic regression:", fbs_logistic4)
        print('\n')

5  vecinos:
Target percentage: 0.1
Accuracy score of a logistic regression: 0.9928875219261359
f1 score of a logistic regression: 0.019867549668874173
fbeta score of a logistic regression: 0.012338640713442196


Target percentage: 0.15
Accuracy score of a logistic regression: 0.990052142153447
f1 score of a logistic regression: 0.05045871559633028
fbeta score of a logistic regression: 0.04499615244421044


Target percentage: 0.2
Accuracy score of a logistic regression: 0.985390585578009
f1 score of a logistic regression: 0.05000000000000001
fbeta score of a logistic regression: 0.06491263305884716


Target percentage: 0.3
Accuracy score of a logistic regression: 0.972727491169474
f1 score of a logistic regression: 0.04380791912384162
fbeta score of a logistic regression: 0.10321515604119173


Target percentage: 0.4
Accuracy score of a logistic regression: 0.954009178941298
f1 score of a logistic regression: 0.043
fbeta score of a logistic regression: 0.16541611121691108


Target percen

En este caso el fbeta score se maximiza (0.2361) cuando se utilizan 5 vecinos más cercanos y un porcentaje de balanceo igual al 50%.

In [71]:
for k in [5,6,7,8,9,10]: #número de vecinos
    print(k,' vecinos:')
    for percentage in [0.1,0.15,0.2,0.3,0.4,0.5]:
        adasyn = ADASYN(percentage,random_state=42, n_neighbors = k)
        X_porct, y_porct = adasyn.fit_resample(X_train_, y_train)
        tree.fit(X_porct, y_porct)
        y_pred_tree = tree.predict(X_test_)
        accuracy_tree4=accuracy_score(y_test, y_pred_tree)
        f1_tree4=f1_score(y_test,y_pred_tree,average='binary')
        fbs_tree4=fbeta_score(y_test,y_pred_tree,average='binary', beta = 10 )
        print('Target percentage:',percentage)
        print("Accuracy score of a tree:",accuracy_tree4)
        print("f1 score of a tree:", f1_tree4)
        print("fbeta score of a tree:", fbs_tree4)
        print('\n')    

5  vecinos:
Target percentage: 0.1
Accuracy score of a tree: 0.9869524473172021
f1 score of a tree: 0.13946117274167988
fbeta score of a tree: 0.17857429880253956


Target percentage: 0.15
Accuracy score of a tree: 0.9826753490160272
f1 score of a tree: 0.05504587155963303
fbeta score of a tree: 0.08477895914941243


Target percentage: 0.2
Accuracy score of a tree: 0.9826513203738857
f1 score of a tree: 0.05497382198952879
fbeta score of a tree: 0.08477557056636956


Target percentage: 0.3
Accuracy score of a tree: 0.9743614388350914
f1 score of a tree: 0.11305070656691606
fbeta score of a tree: 0.26977767303008876


Target percentage: 0.4
Accuracy score of a tree: 0.8972294975610928
f1 score of a tree: 0.03169572107765452
fbeta score of a tree: 0.24658203125


Target percentage: 0.5
Accuracy score of a tree: 0.8569094360477689
f1 score of a tree: 0.027437530622243998
fbeta score of a tree: 0.279281058660873


6  vecinos:
Target percentage: 0.1
Accuracy score of a tree: 0.9903164572170

En este caso el fbeta score se maximiza (0.2793) cuando se utilizan 6 o 10 vecinos más cercanos y un porcentaje de balanceo igual al 50%.

# Exercise 15.6 (3 points)

Compare and comment about the results

1. El árbol reporta mejores resultados y métricas de rendimiento que la regresión logística.
2. Cuando se usó undersampling, el porcentaje de balanceo adecuado fue igual al 15% en ambos casos.
3. Cuando se usó oversampling el porcentaje adecuado fue 20% cuando se corre una regresión logística y 10% cuando se utilizó un árbol.
4. Cuando se usó la función SMOTE, el fbeta score se maximiza cuando se utiliza un porcentaje de balanceo igual al 50%, y 8 vecinos más cercanos para el caso de la regresión logística, y 5 cuando se utiliza un arbol de decisión.  
5. Cuando se usó la función ADASYN, el fbeta score se maximiza cuando se utiliza un porcentaje de balanceo igual al 50%, y 5 vecinos más cercanos para el caso de la regresión logística, y 6 O 10 cuando se utiliza un arbol de decisión.  
6. Cuando se utilizó la función ADASYN, el fbeta score no cambió mucho a medida que cambiaba el número de vecinos más cercanos, manteniendo el porcentaje de balanceo en 50%.
7. Implementar la función ADASYN requiere tiempo de procesamiento.

In [72]:
###############